scoped-mutex/src/lib.rs

```diff
@@ -1,57 +1,100 @@
 #![no_std]
 
 use core::cell::UnsafeCell;
-use core::sync::atomic::{AtomicBool, Ordering};
+
+pub mod impls;
 
 /// Raw mutex trait.
 ///
 /// This mutex is "raw", which means it does not actually contain the protected data, it
-/// just implements the mutex mechanism. For most uses you should use [`super::Mutex`] instead,
-/// which is generic over a RawMutex and contains the protected data.
-///
-/// Note that, unlike other mutexes, implementations only guarantee no
-/// concurrent access from other threads: concurrent access from the current
-/// thread is allowed. For example, it's possible to lock the same mutex multiple times reentrantly.
-///
-/// Therefore, locking a `RawMutex` is only enough to guarantee safe shared (`&`) access
-/// to the data, it is not enough to guarantee exclusive (`&mut`) access.
+/// just implements the mutex mechanism. For most uses you should use [`BlockingMutex`]
+/// instead, which is generic over a RawMutex and contains the protected data.
 ///
 /// # Safety
 ///
 /// RawMutex implementations must ensure that, while locked, no other thread can lock
-/// the RawMutex concurrently.
+/// the RawMutex concurrently. This can usually be implemented using an [`AtomicBool`]
+/// to track the "taken" state. See [crate::impls] for examples of correct implementations.
+///
+/// [`AtomicBool`]: core::sync::atomic::AtomicBool
 ///
 /// Unsafe code is allowed to rely on this fact, so incorrect implementations will cause undefined behavior.
-pub unsafe trait RawMutex {
-    /// Create a new `RawMutex` instance.
+///
+/// # Implementation Note:
+///
+/// This is actually a marker trait for types that implement [`UnconstRawMutex`] and
+/// [`ConstInit`]. This is to allow cases where a mutex cannot be created in const
+/// context, for example some runtime/OS mutexes, as well as testing mutexes like
+/// those from `loom`.
+///
+/// If you are implementing your own RawMutex primitive, you should implement the
+/// [`UnconstRawMutex`] and [`ConstInit`] traits, and rely on the blanket impl
+/// of `impl<T: UnconstRawMutex + ConstInit> RawMutex for T {}`.
+pub trait RawMutex: UnconstRawMutex + ConstInit {}
```

Comment:
not totally sure if i think this alias is pulling its weight, personally --- is it really that much easier for implementations to have `T: RawMutex` bounds instead of just `T: UnconstRawMutex + ConstInit`? IMO it doesn't seem super necessary, and i find the name `UnconstRawMutex` kinda unfortunate...

also, it kind of seems like having this will encourage implementations to just bound their raw mutex with `T: RawMutex` and _also_ require the `ConstInit` trait for _all_ their methods, when they don't actually need it for anything except for const constructors. i kind of feel like just having "UnconstRawMutex" be the "RawMutex" trait would make it more likely for people to put the const constructor in a separate `where T: ConstInit` impl block, instead of requiring it for everything when they don't really need to.

i'm open to being convinced otherwise, if you think this is a major advantage for users.